If we temporarily revert the fix in PR #305, the MMSU CI will fail with the following error:
`FAILED tests/test_model/test_qwen3_omni_mmsu_ci.py::test_mmsu_accuracy_and_speed - AssertionError: MMSU had 3/2000 failed requests (timeouts or empty responses); any failure fails the test.`
This indicates that the current test is capable of preventing this issue from recurring.

--- a/benchmarks/README.md
+++ b/benchmarks/README.md
@@ -7,7 +7,7 @@ and accuracy (WER, MMSU, MMMU) across supported modality combinations.
 
 ```
 benchmarks/
-├── tasks/          # Per-task logic (tts, mmsu, visual_understand)
+├── tasks/          # Per-task logic (tts, audio_understanding, visual_understand)
 ├── metrics/        # Metric computation (performance, accuracy)
 ├── dataset/        # Dataset loaders + download helpers
 ├── benchmarker/    # Framework: runner, data structures, utilities

--- a/benchmarks/dataset/mmsu.py
+++ b/benchmarks/dataset/mmsu.py
@@ -61,6 +61,10 @@ def load_mmsu_samples(
     (train split, ~5000 samples).  zhaochenyang20/mmsu-ci-2000 to
     load our pre-built subset for CI.
     """
+    import tempfile
+
+    from datasets import Audio, load_dataset
+
     ds = load_dataset(repo_id or "ddwang2000/MMSU")
     assert list(ds.keys()) == [
         "train"

--- a/benchmarks/dataset/mmsu.py
+++ b/benchmarks/dataset/mmsu.py
@@ -55,20 +55,12 @@ def load_mmsu_samples(
 ) -> list[MmsuSample]:
     """Load MMSU samples.
 
-    Args:
-        max_samples: Cap on how many samples to return.  None = all.
-        task_names: Optional task_name filter.
-        categories: Optional category filter.
-        seed: If set, shuffle with ``random.Random(seed)`` before slicing.
-        repo_id: HuggingFace dataset repo to load from.  Defaults to
-            None which loads the full ddwang2000/MMSU (train split,
-            ~5000 samples).  Pass a repo id like
-            "zhaochenyang20/mmsu-ci-2000" to load a pre-built subset.
+
+    Note (Yifei, Chenyang):
+    repo_id defaults to None which loads the full ddwang2000/MMSU
+    (train split, ~5000 samples).  zhaochenyang20/mmsu-ci-2000 to
+    load our pre-built subset for CI.
     """
-    import tempfile
-
-    from datasets import Audio, load_dataset
-
     ds = load_dataset(repo_id or "ddwang2000/MMSU")
     assert list(ds.keys()) == [
         "train"

--- a/benchmarks/eval/benchmark_omni_mmsu.py
+++ b/benchmarks/eval/benchmark_omni_mmsu.py
@@ -118,7 +118,7 @@ async def run(
             task_names=args.task_names.split(",") if args.task_names else None,
             categories=args.categories.split(",") if args.categories else None,
             seed=args.seed,
-            repo_id=getattr(args, "repo_id", None),
+            repo_id=args.repo_id,
         )
 
     save_audio_dir = None
@@ -162,8 +162,8 @@ async def run(
     if audio_mode:
         wer_results = compute_text_audio_consistency(
             request_results,
-            getattr(args, "lang", "en"),
-            getattr(args, "asr_device", "cuda:0"),
+            args.lang,
+            args.asr_device,
         )
         output["wer"] = wer_results
         print_wer_summary(wer_results["summary"], args.model)

--- a/tests/test_model/test_qwen3_omni_mmsu_ci.py
+++ b/tests/test_model/test_qwen3_omni_mmsu_ci.py
@@ -14,7 +14,6 @@ from __future__ import annotations
 
 import argparse
 import asyncio
-import subprocess
 import sys
 from pathlib import Path
 
@@ -24,6 +23,7 @@ from benchmarks.dataset.prepare import DATASETS
 from benchmarks.eval.benchmark_omni_mmsu import run as run_mmsu
 from sglang_omni.utils import find_available_port
 from tests.utils import (
+    ServerHandle,
     apply_slack,
     assert_speed_thresholds,
     start_server_from_cmd,
@@ -62,8 +62,7 @@ def server_process(tmp_path_factory: pytest.TempPathFactory):
         "qwen3-omni",
     ]
     proc = start_server_from_cmd(cmd, log_file, port, timeout=STARTUP_TIMEOUT)
-    proc.port = port
-    yield proc
+    yield ServerHandle(proc=proc, port=port)
     stop_server(proc)
 
 
@@ -88,12 +87,15 @@ def _build_args(port: int, output_dir: str) -> argparse.Namespace:
         disable_tqdm=True,
         seed=None,
         repo_id=DATASETS["mmsu-ci-2000"],
+        # Unused in text-only mode but kept for API consistency with run().
+        lang="en",
+        asr_device="cuda:0",
     )
 
 
 @pytest.mark.benchmark
 def test_mmsu_accuracy_and_speed(
-    server_process: subprocess.Popen,
+    server_process: ServerHandle,
     tmp_path: Path,
 ) -> None:
     """Run MMSU eval and assert accuracy and speed meet thresholds."""

--- a/tests/test_model/test_qwen3_omni_mmsu_tts_consistency_ci.py
+++ b/tests/test_model/test_qwen3_omni_mmsu_tts_consistency_ci.py
@@ -20,9 +20,8 @@ from __future__ import annotations
 
 import argparse
 import asyncio
-import subprocess
 import sys
-from copy import copy
+from copy import deepcopy
 from pathlib import Path
 
 import pytest
@@ -32,6 +31,7 @@ from benchmarks.dataset.prepare import DATASETS
 from benchmarks.eval.benchmark_omni_mmsu import run as run_mmsu
 from sglang_omni.utils import find_available_port
 from tests.utils import (
+    ServerHandle,
     apply_slack,
     assert_speed_thresholds,
     assert_wer_results,
@@ -98,8 +98,7 @@ def server_process(tmp_path_factory: pytest.TempPathFactory):
         "qwen3-omni",
     ]
     proc = start_server_from_cmd(cmd, log_file, port, timeout=STARTUP_TIMEOUT)
-    proc.port = port
-    yield proc
+    yield ServerHandle(proc=proc, port=port)
     stop_server(proc)
 
 
@@ -130,7 +129,7 @@ def _build_args(port: int, output_dir: str) -> argparse.Namespace:
 
 @pytest.mark.benchmark
 def test_mmsu_audio_wer_and_speed(
-    server_process: subprocess.Popen,
+    server_process: ServerHandle,
     tmp_path: Path,
 ) -> None:
     """Run MMSU eval with audio and assert WER and speed meet thresholds."""
@@ -141,7 +140,7 @@ def test_mmsu_audio_wer_and_speed(
     # audio encoder sees a cached+non-cached mixed batch. Inert at
     # concurrency=1; starts to take effect once concurrency is raised.
     base = load_mmsu_samples(max_samples=MAX_SAMPLES, repo_id=DATASETS["mmsu-ci-2000"])
-    dup = copy(base[0])
+    dup = deepcopy(base[0])
     dup.sample_id = f"{base[0].sample_id}__dup"
     samples = [*base, dup]
 

--- a/tests/utils.py
+++ b/tests/utils.py
@@ -8,12 +8,21 @@ import signal
 import statistics
 import subprocess
 from contextlib import contextmanager
+from dataclasses import dataclass
 from pathlib import Path
 from typing import Generator
 
 from benchmarks.benchmarker.utils import wait_for_service
 
 STARTUP_TIMEOUT = 600
+
+
+@dataclass
+class ServerHandle:
+    """Typed bundle of a running server's process and its port."""
+
+    proc: subprocess.Popen
+    port: int
 
 
 @contextmanager

